src/main.rs

```diff
@@ -155,45 +201,50 @@ fn read_single_source(src: &Src) -> Result<String> {
     Ok(res)
 }
 
-fn reformat_file_in_place(file: &PathBuf) -> Result<()> {
-    let input = fs::read_to_string(file)?;
-    let output = nixpkgs_fmt::reformat_string(&input);
-    if input != output {
-        fs::write(file, &output)?;
-    }
-    Ok(())
-}
-
-fn reformat_dir_in_place(dir: &PathBuf) -> Result<()> {
-    let nix_file_type = {
+fn reformat_dir_in_place(
+    dir: &PathBuf,
+    write_changes: bool,
+    files_changed: &mut Vec<PathBuf>,
+    files_count: &mut u64,
+) -> Result<()> {
+    let nix_file_types = {
         let mut builder = ignore::types::TypesBuilder::new();
         builder.add_defaults();
         builder.add("nix", "*.nix").unwrap();
         builder.select("nix");
         builder.build().unwrap()
     };
-    let has_errors = Arc::new(AtomicBool::new(false));
-    ignore::WalkBuilder::new(dir).threads(8).types(nix_file_type).build_parallel().run(|| {
-        let has_errors = Arc::clone(&has_errors);
-        Box::new(move |entry| match reformat_dir_entry(entry) {
-            Ok(()) => ignore::WalkState::Continue,
+
+    let mut has_errors = false;
+    for entry in ignore::WalkBuilder::new(dir).types(nix_file_types).build() {
```

Comment:
It's a shame to loose parallelism here. I don't think that `ignore` has a straightforward API for aggregation, so fixing it unfortunatelly is not trivial.

Here's I think what would work: 

```
let (sender, receiver) = std::mpsc::unbounded();
ignore::WalkBuilder::new(dir).threads(8).types(nix_file_type).build_parallel().run(move || {
    struct Acc { count: u64, paths: Vec<PathBuf>, sender: mpsc::Sender<(u64, Vec<u64>)>};
    impl Drop for Acc {
        fn drop(&mut self) {
            let paths = std::mem::take(&mut self.paths);
            self.sender.send((self.count, paths)).unwrap();
        }
    }
    let mut acc = Acc { count: 0, paths: vec![], sender: sender.clone() };
    Box::new(move |entry| reformat_dir_entry(entry, &mut acc.count, &mut acc.paths))
})?;

for (count, paths) in receiver {
    // aggregate into passed-in arguments
}
```

That' basically a map-reduce

Reply:
Yeah it's a shame. On my computer reformatting nixpkgs went from 5.5s (master) to 20.1s (this branch). Another issue is that all the touched files are only printed at the end of the evaluation so there isn't much progress going on.

Reply:
Hm, if we want to report progress as we go, the solution would be simpler (although slightly less efficient): just push *each* processed file to a channel.

Reply:
Yeah that's what I ended-up implementing.
* 516ba9d - report each file over a channel
* ec7e496 - restored multi-threading, super simple change

We're back at ~5s.

I also played with `.threads(0)` (the default) and the time was ~10s.